README.md
osagitfilter
============

Filter to put [OSA][] languages (AppleScript, JavaScript) into a `git`-repository. 

Default, it prevents from accidently checkin in AppleScript files with Debugging Mode (from [AppleScript Debugger][asdbg]) switched on.

Based on [this answer by Daniel Trebbien][so-ascr-in-git] on stackoverflow and help from [guys on the Script Debugger Forum][asdbg-forum].



Configuration
-------------

Install the filter by running the following command:

	./setup.sh install


Configure the filter by running [git config][gitconfig] (with or without the `--global` argument):

	git config --global filter.osa.clean "osagitfilter --clean %f"
	git config --global filter.osa.smudge "osagitfilter --smudge %f"
	git config --global filter.osa.required "true"


Put the line `*.scpt filter=osa` in your [gitattributes][], for example by running the following command from within your repository directory:

	echo "*.scpt filter=osa" >> .gitattributes



[OSA]: https://developer.apple.com/library/content/documentation/AppleScript/Conceptual/AppleScriptX/Concepts/osa.html "Apple's Open Scripting Architecture"
[asdbg]: http://latenightsw.com
[so-ascr-in-git]: https://stackoverflow.com/a/14425009/56
[asdbg-forum]: http://forum.latenightsw.com/t/cross-play-between-script-debugger-and-script-editor/834/5
[gitconfig]: https://git-scm.com/docs/git-config
[gitattributes]: https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes


Developer info
--------------

Because of the test files, don't enable `osagitfilter` for this `git`-repository.


HISTORY.md
Note on history
===============

I wanted to change my work email to private, so I've rebuild the repository ground up. 
I'd already pushed stuff to GitHub, so I recreated the repo their.
Original `git log` was:

	commit 4ac89601de389fb9763e98e741ac6e30fd100bd8 (HEAD -> master, origin/master, origin/HEAD)
	Author: Doeke Zanstra <doeke@archipunt.nl>
	Date:   Fri Dec 29 16:43:28 2017 +0100

	    Bugfixes and added automated tests
    
	    - Version 0.4
	    - Renamed `--opt-out` to `--forbidden`, use `-` to allow all
	    - Re-added `--no-header`
	    - Atomic logging
	    - Logging now call, caller
	    - Renamed `test` folder to `test-files`

	commit 9459b9ad185af14690bbeebaa7bada8244802f79
	Author: Doeke Zanstra <doeke@archipunt.nl>
	Date:   Tue Dec 26 17:06:35 2017 +0100

	    Some polishing
    
	    - added EXIT trap
	    - works now completely from stdin/out
	    - added --no-header option
	    - renamed blacklist to opt-out
	    - script version
	    - working with a temp-directory now
	    - setup: added log rotate option

	commit b2126422d3fe9be4f04c6952e48a301b22994389
	Author: Doeke Zanstra <doeke@archipunt.nl>
	Date:   Tue Dec 26 14:46:43 2017 +0100

	    Fixed some bugs, and added logging

	commit 271e7a9bbc32ea08f2562254850400a79d1c9f04
	Author: Doeke Zanstra <doeke@archipunt.nl>
	Date:   Mon Dec 25 18:09:02 2017 +0100

	    Oops
    
	    Because of `setup.sh`, the `.sh` extensions are not necessary in the git config.

	commit 6222cf5a06d7bb1d2223279d644e794c16c9778b
	Author: Doeke Zanstra <doeke@archipunt.nl>
	Date:   Mon Dec 25 18:07:27 2017 +0100

	    Added setup and test-files

	commit 47f6ca73d57468bbeaa9a072f9dbd55647f9adc1
	Author: Doeke Zanstra <doeke@archipunt.nl>
	Date:   Mon Dec 25 18:05:47 2017 +0100

	    First version of the osagitfilter

	commit bbfd6ff6dadd57b86b41542e177bdde4bcc86af0
	Author: Doeke Zanstra <doeke@archipunt.nl>
	Date:   Mon Dec 25 17:36:17 2017 +0100

	    My initial git-ascr-filter.sh
    
	    Put in for sentimental reasons

	commit f44a9b7a1e49d4a60b6514521accee6c4cada94e
	Author: Doeke Zanstra <doekman@icloud.com>
	Date:   Mon Dec 25 11:43:40 2017 +0100

	    Initial commit


test.sh
#!/usr/bin/env bash


function lang_test {
	((TEST_NR+=1))
	if (( RUN_TEST != 0 && TEST_NR != RUN_TEST )); then
		((TESTS_SKIPPED+=1))
		return
	fi
	#input arguments
	local CMD="osagetlang $TEST_FILES_DIR/$1"
	local EXPECTED_LANG=$2
	local EXPECTED_EXIT_CODE=$3
	local TEST_DESCRIPTION=$4
	#
	local TEST_LOG=$TEST_DIR/$TEST_NR.stderr.log

	echo "|≣≡=- $(printf '%2s' $TEST_NR), running lang-test: $TEST_DESCRIPTION"
	echo "---===[ $CMD ]=========---------" > $TEST_LOG
	RESULT=$(bash -c "$CMD" 2>>$TEST_LOG)
	EXIT_CODE=$?
	echo "-=> EXIT_CODE:$EXIT_CODE" >> $TEST_LOG
	if (( $EXIT_CODE != $EXPECTED_EXIT_CODE )); then
		((TESTS_NOK+=1))
		echo "# Test failed, exit code $EXPECTED_EXIT_CODE expected; was $EXIT_CODE (output was '$RESULT')"
		return
	fi
	if (( $EXIT_CODE == 0 )); then
		if [[ $RESULT != $EXPECTED_LANG ]]; then
			((TESTS_NOK+=1))
			echo "# Output '$EXPECTED_LANG' expected, but actually got '$RESULT'"
			return
		fi
	fi
	((TESTS_OK+=1))
}


function filter_test {
	((TEST_NR+=1))
	if (( RUN_TEST != 0 && TEST_NR != RUN_TEST )); then
		((TESTS_SKIPPED+=1))
		return
	fi
	#input arguments
	#local CMD="$1" #Hmm, command substitution...
	local INPUT_FILE="$TEST_FILES_DIR/$2"
	local EXPECTED_FILE="$TEST_FILES_DIR/$3"
	local EXPECTED_EXIT_CODE="$4"
	local TEST_DESCRIPTION="$5"
	#
	local EXPECTED_FILENAME=$(basename "$EXPECTED_FILE")
	local EXPECTED_EXT="${EXPECTED_FILENAME##*.}"
	local ACTUAL_FILE="$TEST_DIR/$TEST_NR.actual.$EXPECTED_EXT"
	local TEST_LOG=$TEST_DIR/$TEST_NR.stderr.log
	local TEST_CMD="cat $INPUT_FILE | $CMD > $ACTUAL_FILE"
	#
	echo "|≣≡=- $(printf '%2s' $TEST_NR), running filter-test: $TEST_DESCRIPTION"
	echo "---===[ cat $INPUT_FILE | $1 > $ACTUAL_FILE }=========---------" > $TEST_LOG
	bash -c "cat $INPUT_FILE | $1 > $ACTUAL_FILE 2>> $TEST_LOG"
	EXIT_CODE=$?
	echo "-=> EXIT_CODE:$EXIT_CODE" >> $TEST_LOG
	if (( $EXIT_CODE != $EXPECTED_EXIT_CODE )); then
		((TESTS_NOK+=1))
		echo "# Test failed, exit code $EXPECTED_EXIT_CODE expected; was $EXIT_CODE"
		return
	fi
	if (( $EXIT_CODE == 0 )); then
		echo "---===[ diff $EXPECTED_FILE $ACTUAL_FILE ]=========---------" >> $TEST_LOG
		diff $EXPECTED_FILE $ACTUAL_FILE >> $TEST_LOG 2>&1
		EXIT_CODE=$?
		echo "-=> EXIT_CODE:$EXIT_CODE" >> $TEST_LOG
		if (( $EXIT_CODE != 0 )); then
			((TESTS_NOK+=1))
			echo "# diff failed, see $TEST_NR.stderr.log"
			if [[ $EXPECTED_EXT != scpt ]]; then #non-binary file
				echo "- COMPARE: opendiff $EXPECTED_FILE $ACTUAL_FILE"
			fi
			return
		fi
	fi
	((TESTS_OK+=1))
}

function clean_up {
	if (( TESTS_NOK == 0 )); then
		rm -Rf $TEST_DIR
	else
		echo "CLEANUP YOURSELF: rm -Rf $TEST_DIR"
	fi
}

#testing pre-requisites
if [[ -z $(which osagetlang) || -z $(which osagitfilter) ]]; then
	>&2 echo "Test-subjects 'osagetlang' and/or 'osagitfilter' not found in the PATH."
	>&2 echo "Maybe you should run './setup.sh install' first?"
	exit 1
fi

#statistics
((TEST_NR=0))
((TESTS_OK=0))
((TESTS_NOK=0))
((TESTS_SKIPPED=0))
#
((TEST_ERROR=0))
TEST_DIR="$(mktemp -d -t osagitfilter.test.tmp)"
TEST_FILES_DIR="test-files"
trap clean_up EXIT INT HUP TERM

if [[ $1 =~ -?[0-9]+ ]]; then
	if (( $1 == -1 )); then
		((TEST_ERROR=1))
		((RUN_TEST=0))
	elif (( $1 <= -2 )); then
		((TEST_ERROR=2))
		((RUN_TEST=0))
	else
		((RUN_TEST=$1))
	fi
else
	((RUN_TEST=0))
fi

echo "Starting tests... (to run one test: '$(basename $0) TEST_NR'; use -1 to force failure on lang-tests, -2 for filter-tests)"
echo

#--| Determine OSA language tests
lang_test "as.scpt"    "AppleScript"          0 "Get language of an AppleScript file"
lang_test "asdbg.scpt" "AppleScript Debugger" 0 "Get language of an AppleScript Debugger file"
lang_test "js.scpt"    "JavaScript"           0 "Get language of an JavaScript file"
lang_test "perl.scpt"  "Perl"                 1 "Get language of non-existing file"
lang_test ""           ""                     1 "Get language without arguments"

#--| Instrumentarium to generate a test-error.
if (( TEST_ERROR == 1 )); then
	lang_test "as.scpt"    "AdobeScript"          0 "They would like that (test-case designed to fail)"
fi


if (( TESTS_NOK != 0 )); then
	echo "$ Tests for osagetlang failed, no use to continue further tests"
else

	#--| (Wrong) argument tests
	filter_test "osagitfilter"     "as.scpt"               "usage.txt"             1 "No arguments"
	filter_test "osagitfilter -h"  "as.scpt"               "usage.txt"             0 "Show usage screen (--help)" #this test-case will haunt me with every new version

	CMD_CLEAN="osagitfilter --clean --log"
	CMD_CLEAN_NO="osagitfilter --clean --no-header --log"
	CMD_CLEAN_ALL="osagitfilter --clean --forbidden - --log"
	CMD_CLEAN_APPLE="osagitfilter --clean --forbidden 'JavaScript:AppleScript Debugger' --log"
	CMD_SMUDGE="osagitfilter --smudge --log"
	CMD_BOTH="osagitfilter --clean --log | osagitfilter --smudge --log"

	#--| Plain AppleScript tests
	filter_test "$CMD_CLEAN"       "as.scpt"               "as-hdr.applescript"    0 "Clean AppleScript"
	filter_test "$CMD_CLEAN_NO"    "as.scpt"               "as.applescript"        0 "Clean AppleScript (--no-header)"
	filter_test "$CMD_CLEAN_APPLE" "as.scpt"               "as-hdr.applescript"    0 "Deny non-Apple languages: AppleScript"
	filter_test "$CMD_SMUDGE"      "as-hdr.applescript"    "as.scpt"               0 "Smudge AppleScript"
	filter_test "$CMD_SMUDGE"      "as.applescript"        "as.scpt"               0 "Smudge AppleScript (without header)"
	filter_test "$CMD_BOTH"        "as.scpt"               "as.scpt"               0 "Pass through AppleScript"

	#--| ScriptDebugger tests
	filter_test "$CMD_CLEAN"       "asdbg.scpt"            "asdbg-hdr.applescript" 1 "Default Deny: forbidden Debugging Mode switched on"
	filter_test "$CMD_CLEAN_ALL"   "asdbg.scpt"            "asdbg-hdr.applescript" 0 "Allow Debugging Mode switched on"
	filter_test "$CMD_CLEAN_APPLE" "asdbg.scpt"            "asdbg-hdr.applescript" 1 "Deny non-Apple languages: AppleScript Debugger"
	#--|  This can't be tested, because on compile-time, a GUID is inserted in the header. This GUID doesn't seem to be related 
	#--|  to «event asDBDBid»... So no easy validation for this one.
	#--|  Also, when saving with SD, file-size increases big time, so be aware of this when testing.
	#filter_test "$CMD_SMUDGE"      "asdbg-hdr.applescript" "asdbg.scpt"            0 "Smudge AppleScript Debugger"
	#filter_test "$CMD_BOTH"        "asdbg.scpt"            "asdbg.scpt"            0 "Pass through AppleScript Debugger"

	#--| JavaScript tests
	filter_test "$CMD_CLEAN"       "js.scpt"               "js-hdr.javascript"     0 "Clean JavaScript"
	filter_test "$CMD_CLEAN_NO"    "js.scpt"               "js-hdr.javascript"     0 "Clean JavaScript (--no-header)"
	filter_test "$CMD_CLEAN_APPLE" "js.scpt"               "js-hdr.javascript"     1 "Deny non-Apple languages: JavaScript"
	filter_test "$CMD_SMUDGE"      "js-hdr.javascript"     "js.scpt"               0 "Smudge JavaScript"
	filter_test "$CMD_SMUDGE"      "js.javascript"         "js.scpt"               1 "Smudge AppleScript (without header)"
	filter_test "$CMD_BOTH"        "js.scpt"               "js.scpt"               0 "Pass through JavaScript"

	#--| Instrumentarium to generate a test-error.
	if (( TEST_ERROR == 2 )); then
		filter_test "$CMD_BOTH"        "js.scpt"               "as.scpt"               0 "AppleScript != Javascript (test-case designed to fail)"
	fi

fi

echo
echo "$TEST_NR tests have ran, $TESTS_OK tested OK, $TESTS_NOK tested NOK and $TESTS_SKIPPED have been skipped."


setup.sh
#!/usr/bin/env bash

BASE_DIR=$(pwd)
INSTALL_INTO=/usr/local/bin
COMMANDS='osagetlang osagitfilter'

if [[ $1 = install ]]; then
	echo "Trying to install the osagitfilter-commands"
	for CMD in $COMMANDS; do
		if [[ -f $INSTALL_INTO/$CMD ]]; then
			echo "- ERROR: the command '$CMD' is already exists in '$INSTALL_INTO'. No install."
		else
			if ln -s $BASE_DIR/$CMD.sh $INSTALL_INTO/$CMD; then
				echo "- '$CMD' installed"
			else
				echo "- ERROR: couldn't create symbolic link ($?)"
			fi
		fi
	done
elif [[ $1 = reset ]]; then
	for CMD in $COMMANDS; do
		if [[ -h $INSTALL_INTO/$CMD ]]; then
			if rm -f $INSTALL_INTO/$CMD; then
				echo "- '$CMD' reset"
			else
				echo "- ERROR: couldn't remove '$CMD' ($?)"
			fi
		else
			echo "- ERROR: the command '$CMD' is not installed in '$INSTALL_INTO' as symbolic link."
		fi
	done
elif [[ $1 = rotate ]]; then
	LOG_PATH=~/Library/Logs/Catsdeep/
	LOG_NAME=osagitfilter
	ROTATE_STAMP=$(date "+%Y-%m-%dT%H-%M-%S")
	if [[ -d $LOG_PATH ]]; then
		echo "Rotating log (if any) in $LOG_PATH:"
		if [[ -f $LOG_PATH/$LOG_NAME.log ]]; then
			echo "- renaming $LOG_NAME.log -> ${ROTATE_STAMP}_$LOG_NAME.log"
			mv $LOG_PATH/$LOG_NAME.log $LOG_PATH/${ROTATE_STAMP}_$LOG_NAME.log
		fi
	else
		echo "Creating log directory, nothing to rotate"
		mkdir -p $LOG_PATH
	fi
	touch $LOG_PATH/$LOG_NAME.log
else
	echo "usage: $(basename $0) (install|reset|rotate)"
	echo
	echo "install: create symlinks in '$INSTALL_INTO'"
	echo "  reset: remove those symlinks"
	echo " rotate: rename any old logs to something with a timestamp"
	echo
	echo "Installation status:"
	for CMD in $COMMANDS; do
		if [[ -h $INSTALL_INTO/$CMD ]]; then
			echo "- '$CMD' is currently installed"
		else
			echo "- '$CMD' is currently NOT installed"
		fi
	done
fi


osagetlang.sh
#!/usr/bin/env osascript

use AppleScript version "2.4"
use framework "Foundation"
use framework "OSAKit"

on print line_text
	--On some systems, scripting additions contains a definition for NSURL (where XMLLib.osax is included in scripting additions)
	--We need this elsewhere, so scoping like this works.
	using terms from scripting additions
		do shell script "echo " & quoted form of line_text
	end using terms from
end print

on run args
	if (count of args) is not 1 then error "usage: osagetlang.applescript path-to-script-file.scpt" number 1
	set scpt_path to item 1 of args
	set source_nsurl to current application's NSURL's fileURLWithPath:scpt_path
	set the_script to current application's OSAScript's alloc()'s initWithContentsOfURL:source_nsurl |error|:(missing value)
	if the_script is missing value then error "File or contained script is unreachable" number 1
	set osa_lang to the_script's language()'s |name| as text
	print osa_lang
end run


git-ascr-filter.sh
#!/bin/sh

# From: http://stackoverflow.com/a/14425009/56

if [ $# -ne 2 ]; then
	echo "Usage: $0 --clean/--smudge FILE">&2
	exit 1
else
	if [ "$1" = "--clean" ]; then
		function post_process_apple_script {
			# awk: Detect when Script Debugger's Debug-mode is turned on. And trim trailing spaces while we're at it
			awk '
				#First detect events like «event asDBLine»
				$0 ~ /«event asDB[^»]*»/ {
					system("echo Aborted: AppleScript Debugger its Debugging mode is switched on. >&2");
					exit 6;
				}
				#Trim trailing spaces
				{	gsub(/[ ]+$/, "", $0);
					print $0;
				}'
		}
		set -o pipefail
		osadecompile $2 | post_process_apple_script
	elif [ "$1" = "--smudge" ]; then
		TMPFILE=`mktemp -t tempXXXXXX`
		if [ $? -ne 0 ]; then
			echo "Error: \`mktemp' failed to create a temporary file.">&2
			exit 3
		fi
		if ! mv "$TMPFILE" "$TMPFILE.scpt" ; then
			echo "Error: Failed to create a temporary SCPT file.">&2
			rm "$TMPFILE"
			exit 4
		fi
		TMPFILE="$TMPFILE.scpt"
		# Compile the AppleScript source on stdin.
		if ! osacompile -l AppleScript -o "$TMPFILE" ; then
			rm "$TMPFILE"
			exit 5
		fi
		cat "$TMPFILE" && rm "$TMPFILE"
	else
		echo "Error: Unknown mode '$1'">&2
		exit 2
	fi
fi


osagitfilter.sh
#!/usr/bin/env bash

SCRIPT_VER=0.4

function finish {
	#SCRATCH is initialized to a temp directory only at the moment it's needed
	if [[ $SCRATCH ]]; then
		if [[ $DEBUG = 0 ]]; then
			rm -Rf "$SCRATCH"
		else
			debug "stopped: $(date '+%Y-%m-%d %H:%M:%S') (temporary files have been left for inspection)"
			#make log operation "atomic"
			cat $SCRATCH/tmp.log >> $LOG_PATH/$SCRIPT_NAME.log
		fi
	fi
}
trap finish EXIT
################

function usage {
	echo "---=[ $SCRIPT_NAME - v$SCRIPT_VER ]=----------------------------------------------"
	echo "usage: $SCRIPT_NAME --clean [--forbidden FORBIDDEN] [FILE]  #translates OSA script to text, FORBIDDEN is colon-seperated"
	echo "       $SCRIPT_NAME --smudge [FILE]   #translates text to OSA script"
	if [[ $# > 0 ]]; then
		ERROR 1 "$@"
	fi
	exit 0
}

function debug {
	[[ $DEBUG = 1 ]] && >&2 printf "DEBUG: %s\n" "$@"
	[[ $DEBUG = 2 ]] && printf "%s\n" "$@">>$SCRATCH/tmp.log
}

function ERROR {
	ERR_NR=$1
	shift
	>&2 echo "ERROR: $@"
	debug "ERROR($ERR_NR): $@"
	exit $ERR_NR
}

SCRIPT_NAME=$(basename $0 .sh)
CALLED_WITH="$0 $@"
LOG_PATH=~/Library/Logs/Catsdeep/
OSA_GET_LANG_CMD=osagetlang
DEBUG=0
DEFAULT_OSA_LANG=AppleScript
OSA_LANG=$DEFAULT_OSA_LANG
FORBIDDEN_TEXT="AppleScript Debugger" #colon seperated list
WRITE_HEADER=1
while (( $# > 0 )) ; do
  case $1 in
	-c | --clean)  CMD=clean;;
	-s | --smudge) CMD=smudge;;
	-f | --forbidden) [[ $# > 1 ]] || usage "FORBIDDEN argument expected after $1"; FORBIDDEN_TEXT=$2; shift;;
	-n | --no-header) WRITE_HEADER=0;;
	-d | --debug) DEBUG=1;;
	-l | --log) DEBUG=2;;
	-h | -\? | --help) usage;;
	-v | --version) echo $SCRIPT_VER;exit 0;;
	-*) usage "Unrecognized switch '$1'";;
	*) FILE=$1;;
  esac
  shift
done
IFS=:
if [[ $FORBIDDEN_TEXT = "-" ]]; then
	FORBIDDEN=()
else
	FORBIDDEN=($FORBIDDEN_TEXT)
fi
IFS=$' \n\t'

[[ $CMD ]] || usage "No command supplied"
[[ -d "$LOG_PATH" ]] || mkdir -p "$LOG_PATH"
SCRATCH=$(mktemp -d -t osagitfilter.tmp)

debug "---=[ $SCRIPT_NAME - v$SCRIPT_VER ]=----------------------------------------------"
debug "started: $(date '+%Y-%m-%d %H:%M:%S')"
debug "sw_vers: $(sw_vers | tr -d '\t' | tr '\n' ';')"
debug "call: $CALLED_WITH"
debug "caller: $(ps -o args= $PPID)" #see: https://stackoverflow.com/a/26985984/56
debug "scratch: $SCRATCH"
debug "pwd: $(pwd)"
debug "command: $CMD"
debug "filename: '$FILE'"

if [[ $CMD = clean ]]; then

	#Create a temporary file from the stdin, because osadecompile expects a file, and $FILE might not exist on disk
	CLEAN_SCPT_FILE=$SCRATCH/tmp_clean_stdin.scpt
	cat - > $CLEAN_SCPT_FILE

	#determine osa-language of input file
	OSA_LANG=$($OSA_GET_LANG_CMD $CLEAN_SCPT_FILE)
	debug "OSA lang: $OSA_LANG"
	debug "forbidden langs: ${FORBIDDEN[@]}"

	#check if the osa-lang is forbidden
	for BL in "${FORBIDDEN[@]}"; do
		if [[ $BL = $OSA_LANG ]]; then
			ERROR 1 "OSA language '$BL' is forbidden by $SCRIPT_NAME"
		fi
	done

	#write header
	if [[ $WRITE_HEADER = 0 && "$OSA_LANG" = "$DEFAULT_OSA_LANG" ]]; then
		debug "default OSA language is AppleScript: writing no header because of --no-header"
	else
		if [[ $OSA_LANG = "JavaScript" ]]; then
			comment="//"
		else
			comment="#"
		fi
		echo "$comment@osa-lang:$OSA_LANG"
	fi
	
	#decompile to text, and strip tailing whitespace (never newlines, because of $) and finally remove last line
	debug "Starting osadecompile, strip trailing whitespace and remove last line (which is added by osacompile)"
	osadecompile $CLEAN_SCPT_FILE | sed -E 's/[[:space:]]*$//' | sed -e '$ d'

elif [[ $CMD = smudge ]]; then

	#Create a temporary file, for "random access" of stdin
	SMUDGE_TXT_FILE=$SCRATCH/tmp_smudge_stdin.txt
	#Read git's input and store it in the temp-file, except the first line when it's a meta-comment
	FIRST_LINE_READ=
	while IFS= read -r line; do
		if [[ -z $FIRST_LINE_READ ]]; then
			FIRST_LINE_READ=YES
			debug "first line: '$line'"
			if [[ $line =~ ^(#|\/\/)@osa-lang: ]]; then
				OSA_LANG=$(echo $line | sed -E 's/^(#|\/\/)@osa-lang:(.*)$/\2/')
				debug "osa-lang header: '$OSA_LANG'"
				continue
			fi
		fi
		echo "$line" >> $SMUDGE_TXT_FILE
	done < /dev/stdin
	debug "osa-lang: $OSA_LANG"
	#Create a temporary file, for storing output of osacompile
	SMUDGE_SCPT_FILE=$SCRATCH/tmp_smudge_stdout.scpt
	#Perform the compilation
	debug "Starting osacompilation"
	if ! osacompile -l "$OSA_LANG" -o $SMUDGE_SCPT_FILE < $SMUDGE_TXT_FILE ; then
		ERROR 1 "osacompile failed"
	fi
	#Put the output on the stdout
	cat $SMUDGE_SCPT_FILE

else
	usage "unexpected command (SHOULD NOT HAPPEN)"
fi
